feat: Hash based id (#1440)
lingui/js-lingui: website/docs/ref/macro.md

- You don't need to learn ICU MessageFormat syntax. You always use familiar JS and JSX code.
- Components and functions are type checked.

> - Short id generated for your messages

Reviewer: ```suggestion
- Short ID generated for your messages
```

> - Short ID generated for your messages

Reviewer 2: ```suggestion
- Short ID generated for your messages.
```

  message: "Refresh inbox"
})
```
```js
const msg = defineMessage({
  id: "msg.refresh",
  message: "Refresh inbox"
})

// ↓ ↓ ↓ ↓ ↓ ↓

const msg = /*i18n*/{
  id: "msg.refresh",
  message: "Refresh inbox"
}
```
```js
const msg = defineMessage({
  id: "msg.plural",
  message: plural(count, {
    one: "# Message",
    other: "# Messages"
  })
})

// ↓ ↓ ↓ ↓ ↓ ↓

const msg = /*i18n*/{
  id: "msg.plural",
  message: "{count, plural, one {# Message} other {# Messages}}",
  values: { count }
}
```

### Examples of JSX macros

```jsx
<Trans>Attachment {name} saved</Trans>

> // ↓ ↓ ↓ ↓ ↓ ↓,

Reviewer: ```suggestion
// ↓ ↓ ↓ ↓ ↓ ↓

```

*Message Descriptor* is an object with message ID, default message and other parameters. [`i18n._`](/docs/ref/core.md#i18n._) accepts message descriptors and performs translation and formatting:

```ts
type MessageDescriptor = {
   id: string,
   message?: string,
   values?: Record<string, any>,
   comment?: string
}
```

`id` is message ID and the only required parameter. `id` and `message` are extracted to message catalog. Only `id` and `values` are used at runtime, all other attributes are removed from production code for size optimization.

> You don't need to provide ID manually. Macro will automatically create a short id from your message.

Reviewer: ```suggestion
You don't need to provide ID manually. Macro will automatically create a short ID from your message.
```

#### `context`

The same text elements with different context are extracted with different IDs. For example, if the word "right" uses the following two definitions in two different locations, the word is translated differently and merged back into the application as different translation entries.

- correct as in "you are right"
- direction as in "turn right"

> If the same text elements meet the following conditions, the text elements are extracted only once and use the same ID.

Reviewer: When reading the docs, this section doesn't look it belongs here – the `context` property is not metioned in the `MessageDescriptor` type above.

Author: Change it as you wish.

*Message Descriptor* is an object with message ID, default message and other parameters. [`i18n._`](/docs/ref/core.md#i18n._) accepts message descriptors and performs translation and formatting:

```ts
type MessageDescriptor = {
   id: string,
   message?: string,
   values?: Record<string, any>,
   comment?: string
}
```

> `id` is message ID and the only required parameter. `id` and `message` are extracted to message catalog. Only `id` and `values` are used at runtime, all other attributes are removed from production code for size optimization.

Reviewer 2: ```suggestion
`id` is the message ID and the only required parameter. `id` and `message` are extracted to the message catalog. Only `id` and `values` are used at runtime, all other attributes are removed from production code for size optimization.
```

> You don't need to provide ID manually. Macro will automatically create a short ID from your message.

Reviewer 2: ```suggestion
You don't need to provide your ID manually. Macro will automatically create a short ID from your message.
```

#### `context`

> The same text elements with different context are extracted with different IDs. For example, if the word "right" uses the following two definitions in two different locations, the word is translated differently and merged back into the application as different translation entries.

Reviewer 2: ```suggestion
The same text elements with different contexts are extracted with different IDs. For example, if the word "right" uses the following two definitions in two different locations, the word is translated differently and merged back into the application as different translation entries.
```

            other {{numBooks} books and {numArticles} articles}
         }}
      }`,
    values: { numBooks, numArticles }
  })
```

:::tip
This is just an example how macros can be combined to create a complex messages. However, simple is better because in the end it's the translator who's gonna have to translate these long and complex strings.
:::

:::tip
Use `plural` inside [`t`](#t) or [`defineMessage`](#definemessage) macro if you want to add custom `id`, `context` or `comment` for translators.

```js
const message = t({

>    id: "my.custom.id",
>    comment: "My Comment",
>    message: plural(count, {
>      one: "# Book",
>      other: "# Books"
>    })

Reviewer 2: ```suggestion
  id: "my.custom.id",
  comment: "My Comment",
  message: plural(count, {
    one: "# Book",
    other: "# Books"
  })
```

import { i18n } from "@lingui/core"
const message =
  i18n._(/*i18n*/ {
    id: "V8xI3w",
    message: "{count, selectOrdinal, one {#st} two {#nd} few {#rd} other {#th}}",
    values: { count }
  })
```

:::tip
Use `selectOrdinal` inside [`t`](#t) or [`defineMessage`](#definemessage) macro if you want to add custom `id`, `context` or `comment` for translators.

```js
const message = t({

>    id: "my.custom.id",
>    comment: "My Comment",
>    message: selectOrdinal(count, {
>      one: "#st",
>      two: "#nd",
>      few: "#rd",
>      other: "#th",
>    })
> })

Reviewer 2: ```suggestion
  id: "my.custom.id",
  comment: "My Comment",
  message: selectOrdinal(count, {
    one: "#st",
    two: "#nd",
    few: "#rd",
    other: "#th",
  })
})
```

// ↓ ↓ ↓ ↓ ↓ ↓

import { i18n } from "@lingui/core"

const message =
  i18n._(/*i18n*/ {
    id: "VRptzI",
    message: "{gender, select, male {he} female {she} other {they}}",
    values: { gender }
  })
```

:::tip
Use `select` inside [`t`](#t) or [`defineMessage`](#definemessage) macro if you want to add custom `id`, `context` or `comment` for translators.

```js
const message = t({

>    id: "my.custom.id",
>    comment: "My Comment",
>    message: select(gender, {
>      male: "he",
>      female: "she",
>      other: "they"
>    })

Reviewer 2: ```suggestion
  id: "my.custom.id",
  comment: "My Comment",
  message: select(gender, {
    male: "he",
    female: "she",
    other: "they"
  })
```

Unlike the other JS macros, it doesn't wrap generated *MessageDescription* into [`i18n._`](/docs/ref/core.md#i18n._) call.

```ts
type MessageDescriptor = {
  id?: string,
  message?: string,
  comment?: string,
  context?: string,
}

defineMessage(message: MessageDescriptor)
```

> Either `id` or `message` property is required. `id` is a custom message id. If it isn't set, the `message` is used for generating an id.

Reviewer: ```suggestion
Either `id` or `message` property is required. `id` is a custom message ID. If it isn't set, the `message` (and `context` if provided) are used for generating an ID.
```

// ↓ ↓ ↓ ↓ ↓ ↓

const message = /*i18n*/{
  id: "msg.navigation.about"
}
```

`message` and `comment` are used in message catalogs only.

> `context` is used only for id generation and stripped from output.

Reviewer: ```suggestion
`context` is used only for generating ID and is stripped from the output.
```

:::

## JSX Macros

### `Trans`

| Prop name | Type   | Description                                        |
|-----------| ------ |----------------------------------------------------|
| `id`      | string | Custom message ID                                  |
| `comment` | string | Comment for translators                            |
| `context` | string | Provide different translation for the same message |

[`Trans`](/docs/ref/react.md#trans) is the basic macro for static messages, messages with variables, but also for messages with inline markup:

#### `id`

Each message in catalog is identified by **message ID**.

> While macro use message to generate ID, it's possible to override it.

Reviewer: ```suggestion
While macro uses message (and `context` property if provided) to generate ID, it's possible to override it.
```

```jsx
import { Trans } from "@lingui/macro";
<Trans id="message.attachment_saved">Attachment {name} saved.</Trans>;

> // ↓ ↓ ↓ ↓ ↓ ↓

Reviewer: ```suggestion
// ↓ ↓ ↓ ↓ ↓ ↓

```

Render prop function used to render translation. This prop is directly passed to [`Trans`](/docs/ref/react.md#trans) component from [`@lingui/react`](/docs/ref/react.md). See [rendering of translations](/docs/ref/react.md#rendering-translations) for more info.

#### `context`

The same text elements with different context are extracted with different IDs. For example, if the word "right" uses the following two definitions in two different locations, the word is translated differently and merged back into the application as different translation entries.

- correct as in "you are right"
- direction as in "turn right"

If the same text elements meet the following conditions, the text elements are extracted only once and use the same ID.

```jsx
import { Trans } from "@lingui/macro";
<Trans>Refresh inbox</Trans>;

> // ↓ ↓ ↓ ↓ ↓ ↓

Reviewer: ```suggestion
// ↓ ↓ ↓ ↓ ↓ ↓

```

import { Trans } from "@lingui/react";
<Trans id={"EsCV2T"} message="Refresh inbox" />;
```

> Generates different ID when `context` is provided:

Reviewer: ```suggestion
Lingui generates different IDs when `context` is provided:
```

```jsx
import { Trans } from "@lingui/macro";
<Trans context="direction">right</Trans>;
<Trans context="correctness">right</Trans>;

> // ↓ ↓ ↓ ↓ ↓ ↓

Reviewer: ```suggestion
// ↓ ↓ ↓ ↓ ↓ ↓

```

Props of [`Plural`](/docs/ref/macro.md#plural-1) macro are transformed into [`plural`](/docs/ref/message-format.md) format.

```jsx
import { Plural } from "@lingui/macro";
<Plural value={numBooks} one="Book" other="Books" />;

> // ↓ ↓ ↓ ↓ ↓ ↓

Reviewer: ```suggestion
// ↓ ↓ ↓ ↓ ↓ ↓

```

*/
```

> Use `<Plural>` inside `<Trans>` macro if you want to provide `id`, `context` or `comment`

Reviewer: ```suggestion
Use `<Plural>` inside `<Trans>` macro if you want to provide `id`, `context` or `comment`.
```

/>
```

> Use `<SelectOrdinal>` inside `<Trans>` macro if you want to provide `id`, `context` or `comment`

Reviewer: ```suggestion
Use `<SelectOrdinal>` inside `<Trans>` macro if you want to provide `id`, `context` or `comment`.
```

    other="Their book"
/>
```

> Use `<Select>` inside `<Trans>` macro if you want to provide `id`, `context` or `comment`

Reviewer: ```suggestion
Use `<Select>` inside `<Trans>` macro if you want to provide `id`, `context` or `comment`.
```
